pkg/connector/store.go

```diff
@@ -163,43 +238,35 @@ func (*Store) encode(c Connector) ([]byte, error) {
 
 // decode a connector from []byte to Connector. It uses the Builder to
 // initialize the connector making it ready to be used.
-func (s *Store) decode(raw []byte) (Connector, error) {
-	var typedConnector typedJSON
-	err := json.Unmarshal(raw, &typedConnector)
+func (s *Store) decode(raw []byte) (*Instance, error) {
+	conn := &Instance{}
+	err := json.Unmarshal(raw, &conn)
 	if err != nil {
 		return nil, err
 	}
 
-	var conn Connector
-	switch typedConnector.Type {
-	case TypeSource.String():
-		conn, err = s.connBuilder.Build(TypeSource, ProvisionTypeAPI)
-		if err != nil {
-			return nil, err
-		}
-	case TypeDestination.String():
-		conn, err = s.connBuilder.Build(TypeDestination, ProvisionTypeAPI) // ProvisionType will be overwritten
-		if err != nil {
-			return nil, err
+	if conn.State != nil {
+		switch conn.Type {
+		case TypeSource:
+			var state SourceState
+			stateJSON, _ := json.Marshal(conn.State)
+			err := json.Unmarshal(stateJSON, &state)
+			if err != nil {
+				return nil, err
+			}
+			conn.State = state
```

Comment:
Right, I'll add a comment here 👍 For future reference - the issue is that `Instance.State` is of type `any` so when unmarshalling a JSON string that field will be populated with a `map[string]any`. Once we know the type of the connector we know the type of the state, so at that point the simplest thing to do is to just marshal the state back into JSON (this can't really fail because we actually got it from a JSON, that's why we ignore the error) and then unmarshal it back into the proper state type.

Reply:
Thanks for the explanation, it helps!